src/vocab_tree/vocab_tree.cpp

```diff
@@ -42,36 +42,38 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "vocab_tree.h"
 #include "clustering.h"
 
+#include <Eigen/Dense>
+
 using std::cout;
 using std::endl;
 
-float l2sq(DTYPE *a, DTYPE *b, int dim)
```

Comment:
Please make the comparison in release, since compiler optimization can really change the things.
You can also tweak GCC/clang compilation optimization in the cmake command line:
http://wiki.flightgear.org/Talk:Building_using_CMake